jetty-servlet/src/main/java/org/eclipse/jetty/servlet/ServletHolder.java

```diff
@@ -1229,36 +1244,48 @@ public UnavailableServlet(UnavailableException unavailableException, Servlet ser
             }
             else
             {
-                _servlet = servlet;
-                _available = System.nanoTime() + TimeUnit.SECONDS.toNanos(unavailableException.getUnavailableSeconds());
+                _unavailableServlet = servlet;
+                _available.set(System.nanoTime() + TimeUnit.SECONDS.toNanos(unavailableException.getUnavailableSeconds()));
             }
         }
 
         @Override
         public void service(ServletRequest req, ServletResponse res) throws ServletException, IOException
         {
-            if (_available == -1)
-                ((HttpServletResponse)res).sendError(HttpServletResponse.SC_NOT_FOUND);
-            else if (System.nanoTime() < _available)
-                ((HttpServletResponse)res).sendError(HttpServletResponse.SC_SERVICE_UNAVAILABLE);
-            else
+            while (true)
             {
-                synchronized (ServletHolder.this)
+                long available = _available.get();
+                if (available < 0)
```

Comment:
This doesn't take into account the potential numerical overflow of the timeout timestamp.